### PR TITLE
[PyTorch] Debug dtype casting in operation-based API

### PR DESCRIPTION
**Pull request by timmoon10**
# Description

Some users initialize model dtypes using module casting functions like [`torch.nn.Module.float`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.float). However, we experience weird errors when casting fusible operations, largely because the module casting functions don't properly handle `Float8Tensor`. This PR monkey-patches some module casting functions (`float`, `half`, `bfloat16`) so they properly handle `Float8Tensor`. This PR also adds logic in the linear operation to respect [PyTorch autocasting](https://pytorch.org/docs/stable/amp.html).

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

- Support setting `Float8Tensor.data` to arbitrary tensors
- Monkey-patch module dtype cast functions (e.g. `torch.nn.Module.float`) to handle `Float8Tensor`
- Respect PyTorch autocast dtype in linear operation

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/test_float8tensor.py [hunk 1]
```diff
@@ -351,3 +351,47 @@ def test_serialization(
...
+        assert x.device == y.device
+
+        # Copy from Float8Tensor
+        x0 = Float8Tensor.to_float8(torch.zeros((4, 3, 1), dtype=torch.float32))
+        x.data = x0
+        assert isinstance(x, Float8Tensor)
+        assert x0.size() == x.size() == x._data.size()
+        assert x0.dtype == x.dtype
+        assert x0.device == x.device == x._data.device
+        assert x0._data is x._data
```
Comment (ptrendx) [inline tests/pytorch/test_float8tensor.py:391]: Hmmm. If that is the case then the comment saying it is a "copy" is wrong, right?

**Review by ptrendx — APPROVED** [review]

File: tests/pytorch/test_float8tensor.py [hunk 2]
```diff
@@ -351,3 +351,47 @@ def test_serialization(
...
+        assert x0.size() == x.size() == x._data.size()
+        assert x.dtype == torch.float32
+        assert x.is_cuda and x._data.is_cuda
+        y = x.dequantize()
+        assert not isinstance(y, Float8Tensor)
+        assert x.size() == y.size()
+        assert x.dtype == y.dtype
+        assert x.device == y.device
+
+        # Copy from plain tensor
```
Comment (timmoon10) [inline tests/pytorch/test_float8tensor.py:371]:
```suggestion
        # Set data to plain tensor
```

File: tests/pytorch/test_float8tensor.py [hunk 3]
```diff
@@ -351,3 +351,47 @@ def test_serialization(
...
+        assert x0.size() == x.size() == x._data.size()
+        assert x0.dtype == x.dtype
+        assert x0.device == x.device == x._data.device
+        y = x.dequantize()
+        assert not isinstance(y, Float8Tensor)
+        assert x.size() == y.size()
+        assert x.dtype == y.dtype
+        assert x.device == y.device
+
+        # Copy from Float8Tensor
```
Comment (timmoon10) [inline tests/pytorch/test_float8tensor.py:384]:
```suggestion
        # Set data to Float8Tensor
```

File: tests/pytorch/test_float8tensor.py (hunk 1 above)
Comment (timmoon10) [reply tests/pytorch/test_float8tensor.py:391]: Good point, I've tweaked the comment wording.

**Pull request merged**